### PR TITLE
board: Google Twinkie V2

### PR DESCRIPTION
**Pull request by ualbertagreen**
This is a new board for the google Twinkie V2 tool.

Signed-off-by: Jason Yuan <jasonyuan@google.com>

### CODE REVIEW

File: samples/boards/google_pda/src/crc32.c
Comment (jackrosenthal) [inline samples/boards/google_pda/src/crc32.c:73]: Zephyr already provides a CRC32 implementation: https://docs.zephyrproject.org/apidoc/latest/group__crc.html

Why does this need duplicated?

File: samples/boards/google_pda/src/shell.c
Comment (jackrosenthal) [inline samples/boards/google_pda/src/shell.c:41]: Take a look at the style guide ... `else` must be cuddled. `clang-format` can fix this automatically for you.

File: boards/arm/google_twinkie_v2/Kconfig.defconfig
Comment (jackrosenthal) [inline boards/arm/google_twinkie_v2/Kconfig.defconfig:6]: Please split the board support and the sample application into separate commits.

File: samples/boards/google_pda/src/view.c
Comment (jackrosenthal) [inline samples/boards/google_pda/src/view.c:344]: see the style guide, space is always required after `while`.

`clang-format` can fix this automatically for you.

File: samples/boards/google_pda/src/view.c
Comment (jackrosenthal) [inline samples/boards/google_pda/src/view.c:307]: roll into a loop for each led

File: samples/boards/google_pda/src/view.c
Comment (jackrosenthal) [inline samples/boards/google_pda/src/view.c:322]: same here, this should be a loop

**Review by jackrosenthal — CHANGES_REQUESTED** [review]

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:44]: These are not LEDs, they should not be using the `gpio-leds` compatible. Not sure how the application code looks like but I imagine it's abusing the `gpio-leds` binding and potentially the LED subsystem and APIs.

If the code using this one is application specific (think it is), this node should live in an `app.overlay` file with the application, and have its own binding, which should also be in the application code base.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:65]: Can you name this to match something in the schematics? Maybe `plug_vbus` or `vbus_read_buf`.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:72]: Same as the for the one above and below, use something meaningful from the schematic either for the node name or both node and label.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:62]: Well that does not divide much. :-)

This is from a current sense amplifier, I guess you have the conversion hardcoded in the app. How about creating a compatible for current sense devices and have the shunt value and diff amp gain as properties instead so you can have the output in amps? That'd be cool to have upstream.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:183]: nit: This looks a bit odd to me, I'd rather see the properties before the child nodes, that seems to be common practice around. Can you swap them?

File: boards/arm/google_twinkie_v2/google_twinkie_v2.yaml
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.yaml:2]: Should probably match what's in devicetree "model"... pick one. :-)

File: boards/arm/google_twinkie_v2/google_twinkie_v2_defconfig
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2_defconfig:10]: You probably want to enable the USB stack somewhere around here since you have that as default for shell and console, right now the shell sample fails to build.

File: dts/bindings/adc/shunt-sensor.yaml
Comment (fabiobaltieri) [inline dts/bindings/adc/shunt-sensor.yaml:8]: Can you work on a separate PR for this?  Keeping things isolate to help out the review process.

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]
Well I guess we are waiting for https://github.com/zephyrproject-rtos/zephyr/pull/54634 for the shut commit, should defo not be here.

In the meantime, the board should get some basic documentation as well, just realized it's missing, could you add some? (look for other boards as reference)

**Comment by ualbertagreen** [comment]
> Well I guess we are waiting for #54634 for the shut commit, should defo not be here.
> 
> In the meantime, the board should get some basic documentation as well, just realized it's missing, could you add some? (look for other boards as reference)

@fabiobaltieri @galak  I've moved #54634 in here as a stacked PR to avoid potential confusion. They are still kept as separate commits with separate commit messages so everything's compartmentalized.

I'll also add another commit stacked on top of this PR with a small app in `samples/` that simply reads the voltage/current sensors and toggles the LED to show an example of how the sensor binding and board devicetree files are to be used. That way we can keep the full Twinkie app as a separate repository. Does that sound good with everyone?

**Review by galak — DISMISSED** [review]
you are missing board docs.  please add.

**Comment by galak** [comment]
> > 
> 
> @fabiobaltieri @galak I've moved #54634 in here as a stacked PR to avoid potential confusion. They are still kept as separate commits with separate commit messages so everything's compartmentalized.
> 
> I'll also add another commit stacked on top of this PR with a small app in `samples/` that simply reads the voltage/current sensors and toggles the LED to show an example of how the sensor binding and board devicetree files are to be used. That way we can keep the full Twinkie app as a separate repository. Does that sound good with everyone?

I'm still a little concerned about the code utilizing the shunt binding only being in a sample.

**Comment by fabiobaltieri** [comment]
> I'm still a little concerned about the code utilizing the shunt binding only being in a sample.

There's an `adc_emul` driver, it's already used to test the adc APIs (https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/adc/adc_api/boards/native_posix.overlay), may as well be a test, and probably do `voltage-divider` as well.

File: include/zephyr/drivers/adc/adc_current_sense_shunt.h
Comment (fabiobaltieri) [inline include/zephyr/drivers/adc/adc_current_sense_shunt.h:69]: I would live this out of this header file, there's few ways of setting up the adc channels depending on what the hw supports, just keep the struct, initializer and scaling function here.

File: include/zephyr/drivers/adc/adc_current_sense_shunt.h
Comment (fabiobaltieri) [inline include/zephyr/drivers/adc/adc_current_sense_shunt.h:12]: Looking at other structures we have like `adc_dt_spec`, `gpio_dt_spec`, `pwm_dt_spec`. Can you follow that pattern? This could be `current_sense_dt_spec`. Same for the initializer (`ADC_DT_SPEC_GET`, `ADC_DT_SPEC_GET_BY_IDX` for reference)

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]
Hey this is cool, thans for following up on it!

I'd ask you to restructure the PR (again) for reviewability though as it's mixing board definition, board specific sample and a generic binding. This is mysuggestion:

Split out a separate PR with commits for:
- add the current_sense binding AND associated header with just the struct, initializer(s), scaling function
- add a header with struct and initializer for the existing voltage-divider binding (you may as well, should be pretty much the same stuff)
- add a new test to show how it's used and test it... the testing logic is obviously going to be very simple but there's value in the end-to-end example, you can take https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/input/gpio_keys as reference: it's alredy setup to only run on `native_posix`,  has an overlay file where you could setup a shunt and divider node and then in the test source you can instantiate the corresponding structures, fire up a scale request and make sure it's what you expect, the Linux code that handles this stuff is called `iio-rescale.c` so I'd suggest calling the test something with "rescale" as well and use it to group these helpers.

That would allow reving that part with the DT and ADC maintainers, then the board can be done separately.

**Comment by fabiobaltieri** [comment]
Blocked by https://github.com/zephyrproject-rtos/zephyr/pull/55948 I guess.

**Comment by ualbertagreen** [comment]
Hi @fabiobaltieri @galak @jackrosenthal, the sensor PR has finally been merged. Could you take a look again at this PR. Just a reminder since it's been a while, this is the PR for the Twinkie V2 board and a sample code demonstrating its ability to snoop the usb-c lines. Thanks.

File: boards/arm/google_twinkie_v2/board.cmake
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/board.cmake:3]: Could you keep the `board_runner_args` at the top and `include` at the bottom? Not sure why it's done that way for the other boards but it seems like that's the case for pretty much all of them, let's keep it coherent unless we have a good reason not to.

File: boards/arm/google_twinkie_v2/doc/index.rst
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/doc/index.rst:12]: Don't think it's right to call it an EC in this case, the "EC" term is somewhat specific of the support chip in laptops, it's just the MCU that the board runs. Also it's fairly given since one person is reading this in the Zephyr documentation. :-) I'd say just drop this line.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:36]: This is a bit confusing since there's no actual button, maybe add a note mentioning that this goes to a TP5?

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:72]: nit: drop one blank line

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:145]: move these two after `pinctrl-0`? at least `pinctrl-names` since the `-names` property maps to the `-<n>` ones, makes sense to keep them close to each other.

File: boards/arm/google_twinkie_v2/google_twinkie_v2.dts
Comment (fabiobaltieri) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.dts:179]: nit: swap these two? normally you'd define first the `-<n>` properties and then the names. (at least that's what I'm used to seeing)

File: samples/boards/google_twinkie_v2_pda/src/view.c
Comment (fabiobaltieri) [inline samples/boards/google_twinkie_v2_pda/src/view.c:22]: How about using the LED APIs instead? Add `CONFIG_LED=y`, then you'd only have to get a device for the `gpio-leds` node and you could use the APIs from `zephyr/drivers/led.h` and drop the custom initialization code.

**Review by fabiobaltieri — COMMENTED** [review]
Thanks for your patience, just few nitpicks, mostly cosmetic.

**Comment by fabiobaltieri** [comment]
Heya, I think you accidentally pushed some changes of the first commit into the second one.

**Comment by ualbertagreen** [comment]
> Heya, I think you accidentally pushed some changes of the first commit into the second one.

Sorry about that, it should be fixed now.

**Comment by fabiobaltieri** [comment]
> > Heya, I think you accidentally pushed some changes of the first commit into the second one.
> 
> Sorry about that, it should be fixed now.

Yup no prob, happens to me all the time as well. There's still a couple of unresolved comments, all nitpicky, doule blank lines, odd property ordering etc, could you have a look at those? Happy with it after those.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by jackrosenthal — DISMISSED** [review]

File: boards/arm/google_twinkie_v2/google_twinkie_v2.yaml
Comment (str4t0m) [inline boards/arm/google_twinkie_v2/google_twinkie_v2.yaml:9]: The RAM/flash  size look too small, shouldn't they be 144k/512k?

**Review by str4t0m — DISMISSED** [review]

File: boards/arm/google_twinkie_v2/google_twinkie_v2.yaml
Comment (ualbertagreen) [reply boards/arm/google_twinkie_v2/google_twinkie_v2.yaml]: Yes, It has now been corrected.

**Review by fabiobaltieri — APPROVED** [review]

**Review by sambhurst — APPROVED** [review]

**Review by jackrosenthal — APPROVED** [review]

**Pull request merged**
